xlators/mgmt/glusterd/src/glusterd-pmap.c

```diff
@@ -89,337 +87,32 @@ pmap_registry_get(xlator_t *this)
- * the reason for the former is to ensure we don't end up with stale ports
- */
-int
-pmap_assign_port(xlator_t *this, int old_port, const char *path)
-{
-    int ret = -1;
-    int new_port = 0;
-
-    if (old_port) {
-        ret = pmap_registry_remove(this, 0, path, GF_PMAP_PORT_BRICKSERVER,
-                                   NULL, _gf_false);
-        if (ret) {
-            gf_msg(this->name, GF_LOG_WARNING, GD_MSG_PMAP_REGISTRY_REMOVE_FAIL,
-                   0,
-                   "Failed to"
-                   "remove pmap registry for older signin for path"
-                   " %s",
-                   path);
-        }
-    }
-    new_port = pmap_registry_alloc(this);
-    return new_port;
-}
-
-int
-pmap_registry_bind(xlator_t *this, int port, const char *brickname,
-                   gf_pmap_port_type_t type, void *xprt)
-{
-    struct pmap_registry *pmap = NULL;
-    int p = 0;
-
-    pmap = pmap_registry_get(this);
-
-    if (port > pmap->max_port)
-        goto out;
-
-    p = port;
-    if (pmap->ports[p].type == GF_PMAP_PORT_FREE) {
-        /* Because of some crazy race in volume start code path because
-         * of friend handshaking with volumes with quorum enabled we
-         * might end up into a situation where glusterd would start a
-         * brick and get a disconnect and then immediately try to start
-         * the same brick instance based on another friend update
-         * request. And then if for the very first brick even if the
-         * process doesn't come up at the end sign in event gets sent
-         * and we end up having two duplicate portmap entries for the
-         * same brick. Since in brick start we mark the previous port as
-         * free, its better to consider a sign in request as no op if
-         * the corresponding port type is marked as free
-         */
-        goto out;
-    }
-    if (pmap->ports[p].brickname) {
-        char *tmp = pmap->ports[p].brickname;
-        asprintf(&pmap->ports[p].brickname, "%s %s", tmp, brickname);
-        free(tmp);
-    } else {
-        pmap->ports[p].brickname = strdup(brickname);
-    }
-    pmap->ports[p].type = type;
-    pmap->ports[p].xprt = xprt;
-
-    gf_msg("pmap", GF_LOG_INFO, 0, GD_MSG_BRICK_ADD,
-           "adding brick %s on port %d", brickname, port);
-
-    if (pmap->last_alloc < p)
-        pmap->last_alloc = p;
-out:
-    return 0;
-}
-
-int
-pmap_registry_extend(xlator_t *this, int port, const char *brickname)
-{
-    struct pmap_registry *pmap = NULL;
-    char *old_bn;
-    char *new_bn;
-    size_t bn_len;
-    char *entry;
-    int found = 0;
-
-    pmap = pmap_registry_get(this);
-
-    if (port > pmap->max_port) {
-        return -1;
-    }
-
-    switch (pmap->ports[port].type) {
-        case GF_PMAP_PORT_LEASED:
-        case GF_PMAP_PORT_BRICKSERVER:
+    while (true) {
+        p = (rand() % (pmap->max_port - pmap->base_port + 1)) + pmap->base_port;
```

Comment:
I agree with this change, but we'll need to document it properly. Previously, bricks used sequential ports starting at 49152. Even though the potential port range is the same, I know that many users' firewall configurations don't allow the full range. We even say something similar in existing documentation:

https://docs.gluster.org/en/latest/Administrator%20Guide/Setting%20Up%20Clients/

> Ensure that TCP and UDP ports 24007 and 24008 are open on all Gluster servers. Apart from these ports, you need to open one port for each brick starting from port 49152 (instead of 24009 onwards as with previous releases). The brick ports assignment scheme is now compliant with IANA guidelines. For example: if you have five bricks, you need to have ports 49152 to 49156 open.

This needs to be changed and clearly said in the release notes to try to minimize connection issues after upgrade.

Reply:
Will send out a PR for the docs too, regarding the update in the ports allocation.

Reply:
I have created an issue for this in the docs repo https://github.com/gluster/glusterdocs/issues/609
And, once the changes are merged will send out a PR for that too.

Reply:
I really like the idea of cleaning up the code around port mapper. 

However, I don't see how making the port selection random is any better than keeping it sequential but without storing stale state. As @xhernandez  has pointed out, many admins only open up as many ports in their firewall as they need. This is why max_port option was introduced recently, admin can communicate through max_port the highest port number they have open in their firewall for gluster brick. 

Comparing the situations:
1. sequential probing and no max_port set by the admin: Lesser probability of glusterd choosing a port that is blocked by the firewall.
2. random probing and no max_port set by the admin: higher probability of glusterd choosing a port that is blocked by the firewall.
3. sequential probing and max_port set by the admin: O(n) operation for glusterd to find a valid port 
4. random probing and max_port set by the admin: O(n) operation for glusterd to find a valid port

If there a use-case I have missed where random probing helps, certainly go ahead with random. Just wanted to highlight the differences. When you make the change in the doc, please explain the use of max_port as a control for firewall range.

Reply:
@raghavendra-talur thank you for reviewing this PR, I had few things like random probing doesn't guarantee a O(n) time, but sequential probing does. 
Random probing may get into a blocked port, but that's not guaranteed. But, if we use sequential approach we will have to go through the brick ports one by one even though they are block, then isn't that costly?

Please correct me if I am wrong somewhere.

Reply:
Yes, random probing might be a little more efficient than sequential. I will leave that to you. It is not a hard requirement.


The magic port concept isn't the same but that does not block this PR either. I will create a new issue with details for that refactor. Please go ahead with this PR.